frame/staking/src/voter_bags.rs

```diff
@@ -0,0 +1,883 @@
+fn current_bag_for<T: Config>(id: &AccountIdOf<T>) -> Option<VoteWeight> {
+	VoterBagFor::<T>::try_get(id).ok()
+}
+
+/// Data structure providing efficient mostly-accurate selection of the top N voters by stake.
+///
+/// It's implemented as a set of linked lists. Each linked list comprises a bag of voters of
+/// arbitrary and unbounded length, all having a vote weight within a particular constant range.
+/// This structure means that voters can be added and removed in `O(1)` time.
+///
+/// Iteration is accomplished by chaining the iteration of each bag, from greatest to least.
+/// While the users within any particular bag are sorted in an entirely arbitrary order, the overall
+/// stake decreases as successive bags are reached. This means that it is valid to truncate
+/// iteration at any desired point; only those voters in the lowest bag (who are known to have
+/// relatively little power to affect the outcome) can be excluded. This satisfies both the desire
+/// for fairness and the requirement for efficiency.
+pub struct VoterList<T: Config>(PhantomData<T>);
+
+impl<T: Config> VoterList<T> {
+	/// Remove all data associated with the voter list from storage.
+	pub fn clear() {
+		crate::VoterCount::<T>::kill();
+		crate::VoterBagFor::<T>::remove_all(None);
+		crate::VoterBags::<T>::remove_all(None);
+		crate::VoterNodes::<T>::remove_all(None);
+	}
+
+	/// Regenerate voter data from the `Nominators` and `Validators` storage items.
+	///
+	/// This is expensive and should only ever be performed during a migration, never during
+	/// consensus.
+	///
+	/// Returns the number of voters migrated.
+	pub fn regenerate() -> u32 {
+		Self::clear();
+
+		let nominators_iter = Nominators::<T>::iter().map(|(id, _)| Voter::nominator(id));
+		let validators_iter = Validators::<T>::iter().map(|(id, _)| Voter::validator(id));
+		let weight_of = Pallet::<T>::weight_of_fn();
+
+		Self::insert_many(
+			nominators_iter.chain(validators_iter),
+			weight_of,
+		)
+	}
+
+	/// Decode the length of the voter list.
+	pub fn decode_len() -> Option<usize> {
+		crate::VoterCount::<T>::try_get().ok().map(|n| n.saturated_into())
+	}
+
+	/// Iterate over all nodes in all bags in the voter list.
+	///
+	/// Full iteration can be expensive; it's recommended to limit the number of items with `.take(n)`.
+	pub fn iter() -> impl Iterator<Item = Node<T>> {
+		// We need a touch of special handling here: because we permit `T::VoterBagThresholds` to
+		// omit the final bound, we need to ensure that we explicitly include that threshold in the
+		// list.
+		//
+		// It's important to retain the ability to omit the final bound because it makes tests much
+		// easier; they can just configure `type VoterBagThresholds = ()`.
+		let thresholds = T::VoterBagThresholds::get();
+		let iter = thresholds.iter().copied();
+		let iter: Box<dyn Iterator<Item = u64>> = if thresholds.last() == Some(&VoteWeight::MAX) {
```

Comment:
this if/else could also be a test-case.

Reply:
We have implicit test cases, in terms of tests which fail when the if/else behavior is not present, as seen [here](https://gitlab.parity.io/parity/substrate/-/jobs/997426); adding the conditional fixed those. Would you like an explicit test?